src/v/cluster/scheduling/leader_balancer.cc

```diff
@@ -0,0 +1,519 @@
+    if (!_raft0->is_leader()) {
+        vlog(clusterlog.debug, "Leadership balancer tick: not leader");
+        if (!_timer.armed()) {
+            _timer.arm(idle_timeout);
+        }
+        _need_controller_refresh = true;
+        co_return ss::stop_iteration::yes;
+    }
+
+    /*
+     * if we are running the rebalancer after having lost leadership, then
+     * inject a barrier into the raft0 group which will cause us to wait
+     * until the controller has replayed its log. if an error occurs, retry
+     * after a short delay.
+     */
+    if (_need_controller_refresh) {
+        auto res = co_await _raft0->linearizable_barrier();
+        if (!res) {
+            vlog(
+              clusterlog.debug,
+              "Leadership balancer tick: failed to wait on controller "
+              "update: {}. Retrying in {} seconds",
+              res.error().message(),
+              std::chrono::duration_cast<std::chrono::seconds>(
+                leader_activation_delay)
+                .count());
+            if (!_timer.armed()) {
+                _timer.arm(leader_activation_delay);
+            }
+            co_return ss::stop_iteration::yes;
+        }
+        _need_controller_refresh = false;
+    }
+
+    if (_as.local().abort_requested()) {
+        co_return ss::stop_iteration::yes;
+    }
+
+    /*
+     * For simplicity the current implementation rebuilds the index on each
+     * rebalancing tick. Testing shows that this takes up to a couple
+     * hundred microseconds for up to 1000s of raft groups. This can be
+     * optimized later by attempting to minimize the number of rebuilds
+     * (e.g. on average little should change between ticks) and bounding the
+     * search for leader moves.
+     */
+    greedy_balanced_shards strategy(build_index());
+
+    if (clusterlog.is_enabled(ss::log_level::trace)) {
+        auto cores = strategy.stats();
+        for (const auto& core : cores) {
+            vlog(
+              clusterlog.trace,
+              "Leadership balancing stats: core {} leaders {}",
+              core.shard,
+              core.leaders);
+        }
+    }
+
+    // gc expired mutes
+    absl::erase_if(
+      _muted, [now = clock_type::now()](auto g) { return now >= g.second; });
+
+    auto error = strategy.error();
+    auto transfer = strategy.find_movement(muted_groups());
+    if (!transfer) {
+        vlog(
+          clusterlog.info,
+          "No leadership balance improvements found with total error {}",
+          error);
+        _timer.arm(idle_timeout);
+        co_return ss::stop_iteration::yes;
+    }
+
+    auto success = co_await do_transfer(*transfer);
+    if (!success) {
+        vlog(
+          clusterlog.info,
+          "Error transferring leadership group {} from {} to {}",
+          transfer->group,
+          transfer->from,
+          transfer->to);
+        _muted.try_emplace(transfer->group, clock_type::now() + mute_timeout);
```

Comment:
A couple of prometheus stats for successful leadership transfers & failed transfers would be useful to enable monitoring for any unexpected thrashing.

Reply:
Agreed